hail/src/main/scala/is/hail/cxx/EmitTriplet.scala

```diff
@@ -94,4 +94,28 @@ case class SparkFunctionContext(sparkEnv: Code, region: EmitRegion)
 
 abstract class ArrayEmitter(val setup: Code, val m: Code, val setupLen: Code, val length: Option[Code], val arrayRegion: EmitRegion) {
   def emit(f: (Code, Code) => Code): Code
+}
+
+class NDArrayLoopEmitter(fb: FunctionBuilder, nd: Variable, nDims: Int, loopBody: Seq[(Variable, Variable)] => Code) {
+
+  def emit(): Code = {
+    emitLoops(nDims)
+  }
+
+  private def emitLoops(dim: Int, idxVars: Seq[(Variable, Variable)] = Seq.empty): Code = {
+    dim match {
+      case 0 => loopBody(idxVars)
+      case _ =>
+        val i = fb.variable("i", "long")
+        val currDim = fb.variable("dim", "int", s"$nd.flags ? $nDims - $dim : $dim - 1")
```

Comment:
you're checking row-major-ness here, right? might be nice to add a function so you don't have to change this if you change the flag format. Or at least check `$nd.flags & 0x1` in anticipation of maybe adding more flags?